include/net/nrf_cloud_codec.h

```diff
@@ -20,6 +23,498 @@ extern "C" {
  * @{
  */
 
+/** @brief Object types for encoding and decoding */
+enum nrf_cloud_obj_type {
+	NRF_CLOUD_OBJ_TYPE__UNDEFINED,
+
+	NRF_CLOUD_OBJ_TYPE_JSON,
+	NRF_CLOUD_OBJ_TYPE_CBOR,
+
+	NRF_CLOUD_OBJ_TYPE__LAST,
+};
+
+/** @brief Encoded data sources */
+enum nrf_cloud_enc_src {
+	/** Encoded data is not set */
+	NRF_CLOUD_ENC_SRC_NONE,
+	/** Encoded data was set by @ref nrf_cloud_obj_cloud_encode */
+	NRF_CLOUD_ENC_SRC_CLOUD_ENCODED,
+	/** Encoded data was set manually or by @ref NRF_CLOUD_OBJ_PRE_ENC_DEFINE */
+	NRF_CLOUD_ENC_SRC_PRE_ENCODED,
+};
+
+/** @brief Object used for building nRF Cloud messages. */
+struct nrf_cloud_obj {
+
+	enum nrf_cloud_obj_type type;
+	union {
+		cJSON *json;
+		/* TODO:
+		 * struct zcbor_state_t cbor;
```

Comment:
I looked at the code the zcbor code generator output, and learned that each specific use case needs one or more "backup states".  The generated code directly allocates on the stack an array of zcbor_state_t.  There are macros `ZCBOR_STATE_D` and `ZCBOR_STATE_E` for encoding and another for decoding which will allocate these for the user, using the formula:
`num_backups + 2`
where 'num_backups` is roughly how complex the overall message is.  See the not especially helpful docs: https://github.com/NordicSemiconductor/zcbor#the-zcbor-state-object.

I wonder if it makes sense to declare here in this union just a pointer to one or more zcbor_state_t.  This could point to a dynamically allocated array that would be freed when `nrf_cloud_obj_free()` is called.